parcels/grid.py

```diff
@@ -98,16 +102,49 @@ class CRectilinearGrid(Structure):
                         ]
 
         # Create and populate the c-struct object
-        cstruct = CRectilinearGrid(self.lon.size, self.lat.size, self.depth.size,
-                                   self.time.size, -1,
-                                   self.lon.ctypes.data_as(POINTER(c_float)),
-                                   self.lat.ctypes.data_as(POINTER(c_float)),
-                                   self.depth.ctypes.data_as(POINTER(c_float)),
-                                   self.time.ctypes.data_as(POINTER(c_double)))
-        return cstruct
+        if not self.cstruct:  # Not to point to the same grid various times if grid in various fields
+            depthSize = self.depth.size if self.gtype == GridCode.RectilinearZGrid else self.depth.shape[2]
+            self.cstruct = CRectilinearGrid(self.lon.size, self.lat.size, depthSize,
+                                            self.time.size, self.z4d,
+                                            self.lon.ctypes.data_as(POINTER(c_float)),
+                                            self.lat.ctypes.data_as(POINTER(c_float)),
+                                            self.depth.ctypes.data_as(POINTER(c_float)),
+                                            self.time.ctypes.data_as(POINTER(c_double)))
+        return self.cstruct
+
+
+class RectilinearZGrid(RectilinearGrid):
+    """Rectilinear Z Grid
+
+    :param name: Name of the grid
+    :param lon: Vector containing the longitude coordinates of the grid
+    :param lat: Vector containing the latitude coordinates of the grid
+    :param depth: Vector containing the vertical coordinates of the grid, which are z-coordinates.
+           The depth of the different layers is thus constant.
+    :param time: Vector containing the time coordinates of the grid
+    :param time_origin: Time origin of the time axis
+    :param mesh: String indicating the type of mesh coordinates and
+           units used during velocity interpolation:
+
+           1. spherical (default): Lat and lon in degree, with a
+              correction for zonal velocity U near the poles.
+           2. flat: No conversion, lat/lon are assumed to be in m.
+    """
+
+    def __init__(self, name, lon, lat, depth=None, time=None, time_origin=0, mesh='flat'):
+        RectilinearGrid.__init__(self, name, lon, lat, time, time_origin, mesh)
+        if isinstance(depth, np.ndarray):
+            assert(len(depth.shape) == 1), 'depth is not a vector'
 
+        self.gtype = GridCode.RectilinearZGrid
+        self.depth = np.zeros(1, dtype=np.float32) if depth is None else depth
+        self.z4d = -1  # only used in RectilinearSGrid
```

Comment:
why not set `self.z4d` to `None` then?

Reply:
It is sent to C code as an int, so I gave an int (1 for z4d and 0 for z3d, and -1 if it's not even used...)